Add pushmsixscript

gbrownmozilla opened the pull request:
Add pushmsixscript, a new script for pushing .msix files to the Microsoft Store.

In local testing, all tests pass via 'tox'.

In local testing, with appropriate store credentials configured and a hand-edited task.json, I have verified that I can run the script to verify an .msix, obtain an access token, and use the access token to list the existing info for the release channel application on the store.

gbrownmozilla committed:
Add pushmsixscript: push to microsoft store

gbrownmozilla committed:
Rebase for py39

gbrownmozilla committed:
microsoft_store.py rewritten for requests
test_microsoft_store.py verifies all internal functions with requests-mock

gbrownmozilla committed:
updated unit tests

gbrownmozilla committed:
use scriptworker_client

gbrownmozilla committed:
major cleanup, re-pin requirements, get all tests passing and improved test coverage

gbrownmozilla committed:
make task schema more like pushapkscript, for now

gbrownmozilla committed:
remove left-over reference to image_setup.sh (not required for pushmsixscript)

gbrownmozilla committed:
fix erroneous path in Dockerfile

escapewindow commented on README.rst:
```diff
@@ -116,6 +116,17 @@ gecko-3-pushflat     pushflat-prod-relengworker-firefoxci-gecko-3
+==================== ============================================
+gecko-1-pushmsix-dev pushmsix-dev-relengworker-firefoxci-gecko-1
+gecko-3-pushmsix     pushmsix-prod-relengworker-firefoxci-gecko-3
+gecko-1-pushmsix     pushmsix-prod-relengworker-firefoxci-gecko-1
```
To create these pools, we'll need to:
- create `project/releng/scriptworker/v2/...` clients in [clients.yml](https://hg.mozilla.org/ci/ci-configuration/file/tip/clients.yml)
  - when these are created, we reset the `accessToken`s manually and record the secrets, generally in cloudops relengworker sops
- also populate any windows store secrets etc, in relengworker sops. If we deal with Github repos at all (partner-repacks?) then we'd need a github oauth token; the level 3 pool will need the CoT private key.
- once those are done we'll add pushmsix to https://github.com/mozilla-services/cloudops-infra/tree/master/projects/relengworker
- deploy and test the pool(s), possibly to dev first, then to prod
- once those things look good, we'll add to https://github.com/mozilla-releng/k8s-autoscale .

All of the above assumes we have the correct `init_worker.sh` and other `docker.d` files; I'll doublecheck those in this review :)

escapewindow commented on pushmsixscript/README.md:
````diff
@@ -0,0 +1,67 @@
+# edit it with your favorite text editor
+```
+
+There are many values to edit. Example values should give you a hint about what to provide. If not, please see [signingscript's README](https://github.com/mozilla-releng/signingscript#config-json) for more details about allowing URLs, or contact the author for other unclear areas.
````
I suspect this is an old URL?

gbrownmozilla replied:
Yes, good catch. Updated here and in pushapkscript.

escapewindow commented on pushmsixscript/SCOPES.md:
```diff
@@ -0,0 +1,24 @@
+  * Used to tell which Microsoft store a given msix should update.
+
+# Scopes
+* `{scope_prefix}:release`
```
Let's add this and `...:beta` to the [cot_restricted_scopes](https://github.com/mozilla-releng/scriptworker/blob/master/src/scriptworker/constants.py#L321-L339) in scriptworker, to avoid someone being able to craft a try push that pushes to the microsoft store.

escapewindow commented on pushmsixscript/docker.d/init_worker.sh:
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
+set -o errexit -o pipefail
```
In this file we'd `test_var_set` the various env vars that Cloudops will populate the k8s pods with.
We create those variable names in cloudops-infra, and we populate the secret values in our relengworker sops.
I imagine we'll need to at least define whatever creds we need to authenticate with the microsoft store.

escapewindow commented on pushmsixscript/docker.d/worker.yml:
```diff
@@ -0,0 +1,17 @@
+request_timeout_seconds: { "$eval": "REQUEST_TIMEOUT_SECONDS" }
+login_url: { "$eval": "LOGIN_URL" }
+token_resource: { "$eval": "TOKEN_RESOURCE" }
+store_url: { "$eval": "STORE_URL" }
```
All of these `$eval` variables would be good to either explicitly set in `init_worker.sh` (for non-secrets) or to pull from the environment in `init_worker.sh` (for secrets).

(For user/pass combinations, one could argue that the username isn't a secret, but having to update the user in one repo and the password in another, with different release flows, is a pain. Rotating an entire sops file full of dozens or hundreds of secrets, and not knowing which username they belong to, is also a massive pain. I've settled on pairing the user/pass combinations together in sops to avoid both of those pain points.)

escapewindow commented on pushmsixscript/examples/config.example.json:
```diff
@@ -0,0 +1,17 @@
+  "application_ids": {
+    "mock": "",
+    "beta": "",
+    "release": ""
```
Depending on our workflow, we may remove beta and release from the dep/dev pools.

escapewindow commented on pushmsixscript/setup.py:
```diff
@@ -0,0 +1,30 @@
+    entry_points={"console_scripts": ["pushmsixscript = pushmsixscript.script:main"]},
+    license="MPL2",
+    install_requires=["scriptworker_client"] + requirements,
+    classifiers=["Programming Language :: Python :: 3.6", "Programming Language :: Python :: 3.7"],
```
Maybe 3.8 and 3.9 now :)
Probably doesn't matter much since we're not pushing to pypi.

escapewindow commented on pushmsixscript/src/pushmsixscript/artifacts.py:
```diff
@@ -0,0 +1,17 @@
+from scriptworker import artifacts
+from scriptworker.utils import get_single_item_from_sequence
```
Ideally we'll get rid of these scriptworker imports at some point.
Doesn't have to be a blocker.

escapewindow commented on pushmsixscript/src/pushmsixscript/script.py:
```diff
@@ -0,0 +1,49 @@
+import logging
+import os
+
+import scriptworker
```
Ideally we get rid of this import :)

escapewindow commented on pushmsixscript/src/pushmsixscript/microsoft_store.py:
```diff
@@ -0,0 +1,151 @@
+        config (config): the scriptworker configuration.
+        msix_file_path (str): The full path to the msix file to upload.
+        channel (str): The Microsoft Store channel.
+    """
```
Elsewhere we use Google-style docstrings https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html . This looks good!

If this goes into a ReadTheDocs type generated code doc, I've found that:
- I tend to have to use double backticks because ReStructuredText:
```
   foo bar ``foobar``
```
- Best practice says we'd have `Returns:` and `Raises:` sections like `Args:`.

However, our docstring usage and style is pretty mixed in scriptworker-scripts, so I'm pretty open to just having things documented. Perhaps someday we'll lint docstrings and generate RtD docs from them.

escapewindow commented on pushmsixscript/src/pushmsixscript/script.py:
```diff
@@ -0,0 +1,49 @@
+def main(config_path=None):
+    # TODO: Cannot yet use scriptworker_client here, as that does not pass the context
+    # to async_main, which still requires context to use scriptworker.artifacts, etc.
+    return scriptworker.client.sync_main(async_main, config_path=config_path, default_config=get_default_config())
```
If this were `[scriptworker_client.]client.sync_main` we could probably get rid of the scriptworker import.

escapewindow commented on pushmsixscript/src/pushmsixscript/task.py:
```diff
@@ -0,0 +1,26 @@
+        raise TaskVerificationError(f"Channel {channel} not allowed, missing scope {scope}")
+
+    if channel not in ALLOWED_CHANNELS:
+        raise TaskVerificationError('Channel "{}" is not allowed. Allowed ones are: {}'.format(channel, ALLOWED_CHANNELS))
```
We could also check if it's in the config `application_ids`, no?

escapewindow commented on pushmsixscript/src/pushmsixscript/microsoft_store.py:
```diff
@@ -0,0 +1,151 @@
+    body = f"grant_type=client_credentials&client_id={client_id}&client_secret={client_secret}&resource={token_resource}"
+    headers = {"Content-Type": "application/x-www-form-urlencoded; charset=utf-8"}
+    response = requests.post(url, body, headers=headers, timeout=config["request_timeout_seconds"])
+    _log_response(response)
```
Have we verified that no sensitive information gets logged? Otherwise we may want to modify what we log / scrub the response.

gbrownmozilla replied:
I reviewed the logging and removed this call in particular.

escapewindow commented on pushmsixscript/src/pushmsixscript/microsoft_store.py:
```diff
@@ -0,0 +1,151 @@
+        if attempts > COMMIT_POLL_MAX_ATTEMPTS:
+            return False
+        attempts += 1
+        time.sleep(COMMIT_POLL_WAIT_SECONDS)
```
Depending on how long we need to wait on this and how flaky it can be, we could go the `notarization_poller` route. Or maybe we just spin up enough pushmsix workers to deal :)

gbrownmozilla replied:
Unfortunately, I don't have any sense of how long it will take yet, since this only happens once we actually commit a new submission. I'm looking into how / when I can do a fuller test.

escapewindow commented on pushmsixscript/src/pushmsixscript/microsoft_store.py:
```diff
@@ -0,0 +1,151 @@
+    url = f"{login_url}/{tenant_id}/oauth2/token"
+    body = f"grant_type=client_credentials&client_id={client_id}&client_secret={client_secret}&resource={token_resource}"
+    headers = {"Content-Type": "application/x-www-form-urlencoded; charset=utf-8"}
+    response = requests.post(url, body, headers=headers, timeout=config["request_timeout_seconds"])
```
All of these `requests.{post,put,get}` calls should probably have some sort of retry wrap around them, no? e.g. [redo](https://pypi.org/project/redo/). Once we put this into production and we're pushing 100+ times per beta or release, there's plenty of opportunity for a network hiccup or Store blip to throw a wrench into the works. Ideally we're ready for it.

We could retry at the request level, or at the function level, or have the entire `push` call retry. Depending on where/how we do this, we may want to make sure that the various network calls are idempotent.

gbrownmozilla replied:
I added a requests-level retry. Hopefully that will be sufficient.

escapewindow commented on pushmsixscript/tests/test_artifacts.py:
```diff
@@ -0,0 +1,27 @@
+import pytest
+from scriptworker import artifacts
```
You know the drill :)

As an aside on tests: 93% coverage, nice! I prefer 100% but we're far from getting there on the scripts repo. Close on the scriptworker repo.

I did see this:

```
ERROR: invocation failed (exit code 1), logfile: /builds/worker/checkouts/src/pushmsixscript/.tox/coveralls/log/coveralls-2.log
[task 2021-11-29T20:39:22.203Z] ================================== log start ===================================
[task 2021-11-29T20:39:22.203Z] Obtaining file:///builds/worker/checkouts/src/pushmsixscript
[task 2021-11-29T20:39:22.203Z]   Installing build dependencies: started
[task 2021-11-29T20:39:22.203Z]   Installing build dependencies: finished with status 'done'
[task 2021-11-29T20:39:22.203Z]   Checking if build backend supports build_editable: started
[task 2021-11-29T20:39:22.203Z]   Checking if build backend supports build_editable: finished with status 'done'
[task 2021-11-29T20:39:22.204Z]   Getting requirements to build wheel: started
[task 2021-11-29T20:39:22.204Z]   Getting requirements to build wheel: finished with status 'done'
[task 2021-11-29T20:39:22.204Z]   Preparing metadata (pyproject.toml): started
[task 2021-11-29T20:39:22.204Z]   Preparing metadata (pyproject.toml): finished with status 'done'
[task 2021-11-29T20:39:22.204Z] Collecting taskcluster
[task 2021-11-29T20:39:22.204Z]   Using cached taskcluster-44.2.2-py3-none-any.whl (135 kB)
[task 2021-11-29T20:39:22.204Z] ERROR: Could not find a version that satisfies the requirement scriptworker-client (from pushmsixscript) (from versions: none)
[task 2021-11-29T20:39:22.204Z] ERROR: No matching distribution found for scriptworker-client
[task 2021-11-29T20:39:22.204Z] 
[task 2021-11-29T20:39:22.204Z] =================================== log end ====================================
[task 2021-11-29T20:39:22.204Z] ___________________________________ summary ____________________________________
[task 2021-11-29T20:39:22.204Z] ERROR:   coveralls: InvocationError for command /builds/worker/checkouts/src/pushmsixscript/.tox/coveralls/bin/python -m pip install --exists-action w -e . (exited with code 1)
[task 2021-11-29T20:39:22.223Z]
```

escapewindow commented on tox.ini:
```diff
@@ -153,6 +155,17 @@ commands =
+changedir = {toxinidir}/pushmsixscript
+commands =
+    tox -e py39,check
+    - tox -e coveralls
```
Ah, probably that wasn't an error because of this minus.

escapewindow reviewed (approved):
So. I have lots of comments, some of them actionable, some informational.

I think we do have to address the env var / secrets stuff before we can roll out, and retries are super important as well. However, I also believe that landing this PR won't break anything, won't cause us to roll anything half-working pools out, and it'll give us a solid base to keep iterating on, so I'm leaning towards landing this, with or without a ton of fixes. And we can address changes in future PRs. (For any real issues that we don't plan on addressing right away, we probably want to file Github issues for, so we don't forget.) What do you think?

Thank you! This is great work.

gbrownmozilla committed:
Updated for review comments:
 - update signingscript links in READMEs
 - add requests retries
 - remove sensitive logging
